pinot-core/src/main/java/org/apache/pinot/core/data/manager/callback/DataManagerCallback.java

```diff
@@ -0,0 +1,106 @@
+   *
+   * In append-tables callback, this method will create a DefaultIndexSegmentCallback
+   * In upsert-tables callback, this method will create a UpsertDataManagerCallbackImpl
+   *
+   * The {@link IndexSegmentCallback} will be used in the constructor of IndexSegment
+   */
+  IndexSegmentCallback getIndexSegmentCallback();
+
+  /**
+   * process the row after transformation in LLRealtimeSegmentDataManager.processStreamEvents(...) method
+   * it happens after the GenericRow has been transformed by RecordTransformer and before it is indexed by
+   * IndexSegmentImpl, to ensure we can provide other necessary data to the segment metadata.
+   *
+   * In append-tables callback, this method will do nothing
+   * In upsert-tables callback, this method will attach the offset object into the GenericRow object.
+   *
+   * The reason we need this particular logic is that in upsert table, we need to add offset data to the physical data
+   * this will help us to apply the update events from key coordinator to upsert table correctly as the offset
+   * is used as the index to identify which record's virtual column we want to update
+   *
+   * @param row the row of newly ingested and transformed data from upstream
+   * @param offset the offset of this particular pinot record
+   */
+  void processTransformedRow(GenericRow row, StreamPartitionMsgOffset offset);
+
+  /**
+   * process the row after indexing in LLRealtimeSegmentDataManager.processStreamEvents(...) method
+   * it happens after the MutableSegmentImpl has done the indexing of the current row in its physical storage
+   *
+   * In append-tables callback, this method will do nothing
+   * In upsert-tables callback, this method will emit an event to the message queue that will deliver the event to
+   * key coordinator.
+   *
+   * This method ensures that we can emit the metadata for an new entry that pinot just indexed to its internal storage
+   * and let key coordinator to be able to consume those events to process the updates correctly
+   *
+   * @param row the row we just index in the current segment
+   * @param offset the offset associated with the current row
+   */
+  void postIndexProcessing(GenericRow row, StreamPartitionMsgOffset offset);
+
+  /**
+   * perform any necessary finalize operation after the consumer loop finished in LLRealtimeSegmentDataManager.consumeLoop(...)
+   * method. It happens after the consumerloop exits the loop and reached the end criteria.
+   *
+   * In append-tables callback, this method will do nothing
+   * In upsert-tables callback, this method will flush the queue producer to ensure all pending messages are deliverd
+   * to the queue between pinot server and pinot key-coordinator
+   *
+   * this method will ensure that pinot server can send all events to key coordinator eventually before a segment
+   * is committed. If this does not happen we might lose data in case of machine failure.
+   */
+  void postConsumeLoop();
```

Comment:
consumeLoop() is an implementation. If we rename the method, and then this API name will have no meaning. Please indicate clearly _when_ you want this callback, and we can decide on a name. Do you want the callback to happen after successful CONSMING to ONLINE state transition (i.e. the completed segment has been loaded in memory, and we are ready to get rid of the CONSUMING segment memory for this segment). Or, do you really want it to be called even if we may catch up later on with some other server because we got a CATCHUP from the controller? (this can be VERY hard to verify correctness, and I would question why that is needed)

Reply:
Update the method name (onConsumptionStoppedOrEndReached) to indicate that this method is supposed to happen during the consumption stop/end-condition reached. Right now it is fine to call this multiple times if we got catch up call. The purpose of the method is indicated at the method doc: (ensure that pinot server can send all events to key coordinator eventually before a segment is committed). Do you think the new naming would be clear enough to a reader or you think we should add more to the method documentation?